src/usePreparedEffect.ts

```diff
@@ -0,0 +1,31 @@
+import { PrepareHookEffect, PrepareHookFunction } from './types';
+import { DependencyList, useEffect } from 'react';
+import { __REACT_PREPARE__ } from './constants';
+
+export interface PreparedEffectOptions {
+  runOnClient?: boolean;
+  awaitImmediately?: boolean;
+}
+
+const usePreparedEffect = (
+  prepareFunction: PrepareHookFunction,
+  deps: DependencyList = [],
+  opts: PreparedEffectOptions = {},
+) => {
+  const { runOnClient = true, awaitImmediately = false } = opts;
+
+  const effect = (): void => {
+    if (runOnClient) {
+      prepareFunction();
+    }
+  };
+
+  (effect as PrepareHookEffect)[__REACT_PREPARE__] = {
+    prepare: prepareFunction,
+    awaitImmediately,
+  };
+
+  useEffect(effect, deps);
```

Comment:
Another question here is this: When react-dom hydrates the component tree, will the `useEffect` "know" that the deps are unchanged, and not run? I would guess no, since the default behavior is that the hook runs _at least once_.
I have not tested this myself though.

Reply:
Ooh, that's a good point. I actually only now realized that the currenty used `prepared` doesn't do that. I think there is a bug on the events-page that makes it do that, but none of the other routes do. I'll definitely look into this.

With this in mind your previous comment makes perfect sense, an empty dep-array will absolutely make it not run on the client if this is fixed 😄

Reply:
Nice. Yeah the event behavior is not intended and has bugged me for a while.

And yeah, I sort of thought about the SSR render as the "first" render, disregarding whether that's on the client or the server. How to preserve the useeffect accross the server-client I have no idea though :shrug:

Reply:
I think if we want to preserve the behavior of `useEffect` and make this hook work well, we probably need some more custom logic and a way to preserve the state from the server. 

Some ideas:

1. We can do it like other SSR libraries (e.g. loadable-components) and output the result/state of `prepare`, send that in the HTML and load it on the client for use in the hooks (determining whether the prepare fn has been run or not etc.)
2. I guess it might also be possible to create a context which is updated while prepare runs, storing the result, and using that result later.

**1** is probably the easiest to implement, and will move the whole store-state & preserve across server-client issue to the library user, but **2** _might_ be more performant/cleaner in the parent code.

Reply:
Ok, I've done some digging, and it turns out that `prepared` also runs the function on the client as well as the server. The reason we don't refetch in `lego-webapp` is that we have a caching system for avoiding sending multiple of the same GET-request. The fetching-behaiviour is identical as far as I can tell when switching to `withPreparedEffect` :)

So, although I agree it could be a useful feature, considering the added complexity, I think we should just leave it as is and allow the user (`lego-webapp`) to implement a system for avoiding running the same code twice.

Reply:
Well yes, but no. With prepared components, we can control whether the prepareFn runs or not using the `componentDidMount` and `componentWillReceiveProps` options. So whatever behavior you are seeing different places in lego-webapp is explicitly set in the use of prepared.

And yes, even though you can control of new data is fetched or not by using cache-time etc. I think that's kinda a hacky solution to a problem that _we_ control and can fix pretty easily.

Also matching the behavior to that of UseEffect I think makes reasoning about the whole data pipeline easier, as well as avoiding having to fight with the library.

Good defaults is always a plus :)

Reply:
Yeah, I think I agree with you. You are right that we have the option to not run in `componentDidMount` with prepared components, but it doesn't keep track of whether or not it was prepared on the server. That functionality could be implemented fairly easily with a `useRef` that keeps track of how many times the `useEffect` is run, and stops it from running on initial mount, but it is not very useful, as we want to have the app work without SSR as well. Which is why `componentDidMount=false` is not used in `lego-webapp`.

Making it keep track of `prepared` state to always intialize only once would be a pretty great feature though, and if it is possible simply using a `Context` it would be very cool. I'm definitely going to spend some time looking into this!